docs/decisions/0001-ci-cd-interface.md

```diff
@@ -0,0 +1,122 @@
+# CI/CD Interface
+
+* Status: proposed
+* Deciders: [list everyone involved in the decision] <!-- optional -->
+* Date: [YYYY-MM-DD when the decision was last updated] <!-- optional -->
+
+Technical Story: Define Makefile interface between infra and application [#105](https://github.com/navapbc/template-infra/issues/105)
+
+## Context and Problem Statement
+
+In order to reuse CI and CD logic for different tech stacks, we need to establish a consistent interface by which different applications can hook into the common CI/CD infrastructure.
+
+## Proposal
+
+Create a `Makefile` in `template-infra` repo that defines the following make targets:
+
+```makefile
+############
+# Validation
+############
```

Comment:
@kyeah this makes me wonder some related questions:

- do you think there should be several CI workflows based on which files changed?
- should each template application also define their own CI workflow, and installing the template application involves copying over the CI workflow from the template application? in that case we should have some sort of naming convention for install, since the template application might call it .github/workflows/ci.yml, but when a project installs the template, they should rename it like `cp template-application-flask/.github/workflows/ci.yml .github/workflows/ci-api.yml` / `cp template-application-nextjs/.github/workflows/ci.yml .github/workflows/ci-web.yml` / `cp template-infra/.github/workflows/ci.yml .github/workflows/ci-infra.yml`

Reply:
> do you think there should be several CI workflows based on which files changed?

yes I think so — this allows us to be targeted about what we're testing for a given PR or commit, and also makes it less confusing in the Actions tab to distinguish between CI runs for different apps/folders.

> should each template application also define their own CI workflow, and installing the template application involves copying over the CI workflow from the template application?

that's a good question! it does seem like the CI workflows should really live with each application template instead of trying to create a generic interface from the template-infra repo. that also makes the template applications testable on their own 😄 

What do you think about naming those files as `ci-api.yml`, `ci-web.yml`, etc. to begin with? I don't think I have a strong preference — I'm also okay keeping them as `ci.yml` or `ci-app.yml` since that wouldn't require any manual copying or renaming if the repo only has one app.

Reply:
> yes I think so — this allows us to be targeted about what we're testing for a given PR or commit, and also makes it less confusing in the Actions tab to distinguish between CI runs for different apps/folders.

👍 

> What do you think about naming those files as ci-api.yml, ci-web.yml, etc. to begin with? I don't think I have a strong preference — I'm also okay keeping them as ci.yml or ci-app.yml since that wouldn't require any manual copying or renaming if the repo only has one app.

Let's go with a hybrid and name them `ci-infra.yml` and `ci-application.yml` to reflect the template, and I think that will cover 90% of the situations we've seen so far. I'm reluctant to go as far as `ci-api` or `ci-web` since `flask` and `nextjs` could be used for either. In the future if a team needs both `template-application-nextjs` and `template-application-flask` then we can deal with the name conflict and have them manually rename as part of installation.

Reply:
@kyeah one more question before I update the tech spec: I think separate CI workflows makes sense.

But what should be do about CD? This conversation makes me think that `template-infra` would includes the infra for terraform backends, and include a library of reusable modules, but not include infra for any applications
- `/infra/bootstrap/account`
- `/infra/modules`
- `/infra/common`
```

and then the `template-application-nextjs` and `template-applicaiton-flask` would each have
```
- `/infra/app/envs/`
- `/infra/app/template/`
```

then in `template-infra` we'd have a single CD workflow, and parameterize it with `app` and `environment` parameters. So the CD would do something like

```
inputs:
  - app
  - environment

jobs:
  deploy:
    steps:
      - name: build
        run: cd ${{ input.app }} && make build
      - name: push image to ecr
      - name: deploy
        run: cd infra/${{ input.app }}/envs/${{ input.environment }} && terraform apply -var image=...
```

and if a project team had multiple apps in one repo, they'd rename the `app` and `infra/app` folders to something else like `api` and `infra/api`

Reply:
hmm scratch that last idea for now. i think splitting infra code across repos doesn't seem maintainable.

Reply:
Ok proposal updated! PTAL!